rcl_yaml_param_parser/src/parse.c

```diff
@@ -616,60 +795,151 @@ rcutils_ret_t parse_key(
       break;
     case MAP_NODE_NAME_LVL:
       {
-        /// Till we get PARAMS_KEY, keep adding to node namespace
-        if (0 != strncmp(PARAMS_KEY, value, strlen(PARAMS_KEY))) {
+        /// Till we get PARAMS_KEY or PARAMS_DESCRIPTORS_KEY, keep adding to node namespace
+        if (0 == strncmp(PARAMS_KEY, value, strlen(PARAMS_KEY))) {
+          if (*is_new_map) {
+            /// The previous key(last name in namespace) was the node name. Remove it
+            /// from the namespace
+            char * node_name_ns = rcutils_strdup(ns_tracker->node_ns, allocator);
+            if (NULL == node_name_ns) {
+              ret = RCUTILS_RET_BAD_ALLOC;
+              break;
+            }
+
+            ret = _validate_name(node_name_ns, allocator);
+            if (RCUTILS_RET_OK != ret) {
+              allocator.deallocate(node_name_ns, allocator.state);
+              break;
+            }
+
+            ret = find_node(node_name_ns, params_st, node_idx);
+            if (RCUTILS_RET_OK != ret) {
+              break;
+            }
+
+            ret = rem_name_from_ns(ns_tracker, NS_TYPE_NODE, allocator);
+            if (RCUTILS_RET_OK != ret) {
+              RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+                "Internal error adding node namespace at line %d", line_num);
+              break;
+            }
+          }
+
+          /// Bump the map level to PARAMS
+          (*map_level)++;
+        } else if (0 == strncmp(PARAMS_DESCRIPTORS_KEY, value, strlen(PARAMS_DESCRIPTORS_KEY))) {
+          if (*is_new_map) {
+            /// The previous key(last name in namespace) was the node name. Remove it
+            /// from the namespace
+            char * node_name_ns = rcutils_strdup(ns_tracker->node_ns, allocator);
+            if (NULL == node_name_ns) {
+              ret = RCUTILS_RET_BAD_ALLOC;
+              break;
+            }
+
+            ret = _validate_name(node_name_ns, allocator);
+            if (RCUTILS_RET_OK != ret) {
+              allocator.deallocate(node_name_ns, allocator.state);
+              break;
+            }
+
+            ret = find_node(node_name_ns, params_st, node_idx);
+            allocator.deallocate(node_name_ns, allocator.state);
+            if (RCUTILS_RET_OK != ret) {
+              break;
+            }
+
+            ret = rem_name_from_ns(ns_tracker, NS_TYPE_NODE, allocator);
+            if (RCUTILS_RET_OK != ret) {
+              RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+                "Internal error adding node namespace at line %d", line_num);
+              break;
+            }
+          }
```

Comment:
@bpwilcox nit: can we reduce code duplication? `map_level` bumps seem to be the only difference between the last two clauses.